Radiofrequency (RF) block

jhillairet opened the pull request:
First implementation of an RF block.

RF block (proposed name: `RFNetwork`) takes as input a [scikit-rf](https://github.com/scikit-rf/scikit-rf/) `Network` object, which is an N-port described in the frequency domain. It converts the N-port into a state space model PRC transfer function using the [scikit-rf Vector Fitting internal functions](https://scikit-rf.readthedocs.io/en/latest/tutorials/VectorFitting.html) and feeds the poles & residues to a PRC transfer function.

- [x] set scikit-rf as an optional dependency to keep pathsim required dependencies limited 
- [x] first implementation and constructor tests for 1 and 2 ports Network
- [x] tests for 1-port network
- [x] tests for 2-port network
- [x] example: frequency response recovered from Spectrum block

ping @Vinc0110

milanofthe commented:
@jhillairet I checked the RFNetwork block out of curiosity and found a typo, 

```python
super().__init__(Poles=vf.poles.real,
                 Residues=vf.residues.reshape(len(vf.poles), ntwk.nports, ntwk.nports),
                 Const=vf.constant_coeff.reshape(ntwk.nports, ntwk.nports))
```

should be 

```python
super().__init__(Poles=vf.poles,
                 Residues=vf.residues.reshape(len(vf.poles), ntwk.nports, ntwk.nports),
                 Const=vf.constant_coeff.reshape(ntwk.nports, ntwk.nports))
```

And then it seems to work!

This is a small example I ran:

```python
import matplotlib.pyplot as plt
import skrf as rf

from pathsim import Simulation, Connection
from pathsim.blocks import *

nw = RFNetwork(rf.data.ring_slot)
sr = SinusoidalSource(frequency=5e9)
sc = Scope(labels=["src", "ring_slot p1", "ring_slot p2"])

sim = Simulation(
    blocks=[sr, nw, sc], 
    connections=[
        Connection(sr, nw, sc[0]),
        Connection(nw[0], sc[1]),
        Connection(nw[1], sc[2]),
        ],
    dt=5e-13
    )

sim.run(1e-9)
sim.plot()

plt.show()
```

jhillairet commented:
Yes indeed, I've fixed it locally but didn't push it. I'm working on tests (when I can..).

By the way, should two state-space having ABCD arrays organized differently (because of poles/residues order passed in a different order) be assumed equal?

jhillairet commented:
OK, I've added a working example, in which we recover the frequency response of vector fitted model of S11 data, by calculating the response to a gaussian pulse using the Spectrum block. 
While it works, not sure if the simulation parameters are properly choosen.

I've also added some tests, but I found that the signs of A and C matrices were different between scikit-rf vector fitting [`_get_ABCDE`  method](https://github.com/scikit-rf/scikit-rf/blob/2fc80001d150e656c747914101240811600d108d/skrf/vectorFitting.py#L1175). I haven't looked yet why.
 
to discuss : I setup scikit-rf as an optional import ; need to improve robustness in case scikit-rf not installed (warnings, etc).

milanofthe commented:
As for the solvers, this is how I would do it with an adaptive timestep solver

```python


from pathsim.solvers import RKBS32

# create the system connections and simulation setup
sim = Simulation(
    blocks=[src, rfntwk, spc],
    connections=[
        Connection(src, rfntwk, spc[0]),
        Connection(rfntwk, spc[1])
    ],
    # dt=1e-13, <- this is not needed then
    tolerance_lte_abs=1e-16, # this is due to the super tiny states
    tolerance_lte_rel=1e-5,  # so error control is dominated by the relative truncation error
    Solver=RKBS32,
)


```

milanofthe commented:
I think it makes sense to have `scikit-rf` as an optional import with a warning or an exception when its not imported and blocks from the `rf` module are used.

Vinc0110 commented:
I'd like to comment that the vector fitting algorithm in scikit-rf is directly calculating the poles and residues. The function `_get_ABCD()` can then convert them into the state-space matrices. The implementation of `_get_ABCD()` is actually quite dirty and certainly not very efficient. I would be more comfortable to directly use the poles and residues.
Moreover, I think there are different ways to form the state-space matrices. Some use real-valued numbers, others use complex values. We should double check the compatibility if we pass those ABCD matrices to pathsim.

milanofthe commented:
@Vinc0110 jep I can confirm that this is messy. I implemented a real valued minimal gilbert realization [here](https://github.com/milanofthe/pathsim/blob/master/src/pathsim/utils/gilbert.py) to get the statespace matrices. Using the poles and residues directly (although possible) is not super straight forward unfortunately.

milanofthe commented:
Yes, but the question is if we want to use pathsim's realization or scikit-rf's realization of statespace matrices. Because the `TransferFunctionPRC` block just converts the rational model to statespace, inheriting from the `StateSpace` block.
